### PR TITLE
fix(logging): log missing frames more concisely

### PR DESCRIPTION
**Pull request by ashdnazg**
-----------------------
**Bugfix:** This PR addresses issue #5879

## Fix Details
Instead of logging an error for every missing frame, log an error just for the first gap in the frames and how many missing frames are there in addition, for example:
```
Sprite "projectile/typhoon": missing frame 2 and 299996 more.
Sprite "projectile/typhoon": missing @2x frames 0-4999 and 294999 more.
```

## Testing Done
Renamed typhoon~... and the @2x to a bunch of scenarios and checked the results

### CODE REVIEW

File: source/ImageSet.cpp [hunk 1]
```diff
@@ -45,6 +47,44 @@ namespace {
...
+		{
+			message += "frame " + to_string(firstGapStart);
+		}
+		else
+		{
+			message += "frames " + to_string(firstGapStart) + "-" + to_string(firstGapEnd);
+		}
+		size_t extraMissing = totalMissing - (firstGapEnd - firstGapStart + 1);
+		if(extraMissing)
+			message += " and " + to_string(extraMissing) + " more";
```
Comment (tehhowch) [inline source/ImageSet.cpp:84]: I think it's fine to not print the exact missing range. The content creator has their filesystem view and can use that to find gaps once they know one exists.

So e.g. "missing frame X (<totalmissing>)" is fine by me.

**Review by tehhowch — COMMENTED** [review]
Thanks for the PR!

I just have some minor nits, a variable rename request, and a request to simplify the output string, which should simplify the code a bunch too.

File: source/ImageSet.cpp [hunk 2]
```diff
@@ -45,6 +47,44 @@ namespace {
 		// are part of the sprite name, not a frame index.
 		return (IsBlend(path[pos]) ? pos : end);
 	}
+
+	// Log an error if frames are missing in one of the paths vectors.
+	void LogIfMissingFrames(const vector<string>& paths, size_t expectedCount, string prefix, bool is2x)
```
[the remark this answers is not in the document]
Comment (ashdnazg) [reply source/ImageSet.cpp:52]: ah yes, I didn't like this either!

File: source/ImageSet.cpp [hunk 3]
```diff
@@ -45,6 +47,44 @@ namespace {
...
+	{
+		size_t firstGapStart = paths.size();
+		size_t firstGapEnd = expectedCount - 1;
+		const size_t totalFrames = min(paths.size(), expectedCount);
+		size_t totalMissing = expectedCount - totalFrames;
+		for(size_t i = 0; i < totalFrames; ++i)
+		{
+			if(paths[i].empty())
+			{
+				totalMissing++;
```
[the remark this answers is not in the document]
Comment (ashdnazg) [reply source/ImageSet.cpp:62]: irrelevant now

File: source/ImageSet.cpp (hunk 1 above)
Comment (ashdnazg) [reply source/ImageSet.cpp:84]: That simplifies the code significnatly indeed

**Comment by ashdnazg** [comment]
pushed a simplified version

**Comment by MCOfficer** [comment]
> ▶ Try to keep lines to about 80 characters in length, and wrap anything over 120 characters.

http://endless-sky.github.io/styleguide/styleguide.xml

File: source/ImageSet.cpp [hunk 4]
```diff
@@ -45,6 +47,21 @@ namespace {
 		// are part of the sprite name, not a frame index.
 		return (IsBlend(path[pos]) ? pos : end);
 	}
+
+	// Log an error if frames are missing in one of the paths vectors.
+	void LogIfMissingFrames(const vector<string>& frameData, size_t expectedCount, string prefix, bool is2x)
```
Comment (tehhowch) [inline source/ImageSet.cpp:52]:
```suggestion
	void LogIfMissingFrames(const vector<string> &frameData, size_t expectedCount, const string &prefix, bool is2x)
```

File: source/ImageSet.cpp [hunk 5]
```diff
@@ -45,6 +47,21 @@ namespace {
...
+	{
+		auto missingLambda = [](const string &s) { return s.empty(); };
```
Comment (tehhowch) [inline source/ImageSet.cpp:54]:
```suggestion
		auto isMissing = [](const string &s) noexcept -> bool { return s.empty(); };
```

File: source/ImageSet.cpp [hunk 6]
```diff
@@ -45,6 +47,21 @@ namespace {
...
+		const auto totalEmpty = count_if(frameData.begin(), frameData.end(), missingLambda);
+		const auto totalMissing = expectedCount - min(frameData.size(), expectedCount) + totalEmpty;
+		if(!totalMissing)
+			return;
+
+		auto firstMissingIt = find_if(frameData.begin(), frameData.end(), missingLambda);
```
Comment (tehhowch) [inline source/ImageSet.cpp:60]: We make fewer trips through if we

1) find pseudo end PE, begin + min(expected, frameData.size)
2) find first missing frame FM in range [begin, pe)
3) fm == pe? Warn if pe != end and it isn't 2x (logic error), but always exit.
4) count missing in [fm, pe) and log it.

What do you think of this?

**Review by tehhowch — COMMENTED** [review]
Looking better 👍

File: source/ImageSet.cpp (hunk 4 above)
Comment (ashdnazg) [reply source/ImageSet.cpp:52]: good catch

File: source/ImageSet.cpp (hunk 6 above)
Comment (ashdnazg) [reply source/ImageSet.cpp:60]: tbh, IMO this is a relatively minor optimisation in non-performance-heavy code, so readability should be prioritised higher.

Comment (ashdnazg) [reply source/ImageSet.cpp:60]: I changed it slightly to make a single pass and still be mostly readable (I think).

**Review by tehhowch — APPROVED** [review]
I fixed up some style nits and avoided counting 2x frames that the game already says it'll ignore:

e.g. if we touch `images/effects/static+10.png` and `images/effects/static+50@2x.png`, we'll get this output:
```
endless-sky (fix-one-simple-trick)$ ./endless-sky -p
Sprite "effect/static": 40 extra frames for the @2x sprite will be ignored.
Sprite "effect/static": missing frame 7 (3 missing in total).
Sprite "effect/static": missing @2x frame 7 (4 missing in total).
Parse completed.
```

**Pull request merged**